### DISCUSSION TITLE
Can we use tree-sitter as a pager?

### DISCUSSION THREAD

**Discussion by Freed-Wu** [tree-sitter/tree-sitter#2707] [Q&A] [answered]
Can we use tree-sitter as a pager like [bat](https://github.com/sharkdp/bat) and [pygmentize](https://pygments.org/)?

**Comment by yehorb** [comment, marked as answer]
Direct answer - no. As it says in the README, "Tree-sitter is a parser generator tool and an incremental parsing library." It is not possible to use the `tree-sitter` application to read an input and provide paged, formatted, or colored output.

If your question is more akin to "Can pager applications such as `bat` or `less` use `tree-sitter` to provide syntax highlighting? - yes, they can, and it is up to the maintainer (or up to you) to add such support.

A quick search for currently available solutions revealed [`syncat`](https://github.com/foxfriends/syncat).

You can also use `nvim` to open files in read-only mode and use it as an ad-hoc pager, but you will not get colored output into a terminal this way.

**Comment by ahlinc** [comment]
Actually it's possible to use it in a pipe with the less util like: `tree-sitter parse ... | less -r` if to write an own wrapper on top of it.

I permanently use it as the next:
![image](https://github.com/tree-sitter/tree-sitter/assets/14666676/5ec0c4cc-fa0a-4d31-b596-eeb36e69f78d)

**Reply by Freed-Wu** [reply]
where is `ts`? I only found `tree-sitter`.

**Reply by sogaiu** [reply]
IIUC, it's likely to be a user-specific alias or wrapper script.

**Reply by Freed-Wu** [reply]
@ahlinc ,  so is it a wrapper or a new program we don't know? What is it?

**Reply by amaanq** [reply]
it's an alias

**Reply by yehorb** [reply]
If I assume that `ts` means `tree-sitter` and `p` means `parse` subcommand, what do `-oc` flags mean? These are not valid flags for `tree-sitter parse`, at least not for the current version...

**Reply by ahlinc** [reply]
It's a custom CLI that isn't upstreamed yet and can be built and installed by `cargo install --git https://github.com/ahlinc/tree-sitter --branch alpha`, after that you may do a copy of that binary somewhere if you need to use original tree-sitter CLI too. Be aware that it may contain bugs that aren't exist in the original CLI.

**Reply by Freed-Wu** [reply]
Good, wait it be upstreamed. :+1:
